### PR TITLE
Fix undefined behaviour problems in  RecoTracker/TkHitPairs

### PR DESCRIPTION
**Pull request by Dr15Jones**
#### PR description:

The following are meant to fix problems seen in the UBSAN IB.
- Avoid dereference of nullptr in LayerHitMapCache
- Avoid making a 0 length VLA in HitPairGeneratorFromLayerPair

#### PR validation:

The code compiles.

### CODE REVIEW

**Comment by Dr15Jones** [comment]
please test

File: RecoTracker/TkHitPairs/interface/LayerHitMapCache.h [hunk 1]
```diff
@@ -37,8 +37,13 @@ class LayerHitMapCache {
         resize(other.theContainer.size());
 
       for (size_t i = 0, size = other.theContainer.size(); i != size; ++i) {
-        assert(get(i) == nullptr);               // We don't want to override any existing value
-        theContainer[i].reset(*(other.get(i)));  // pass by reference to denote that we don't own it
+        assert(get(i) == nullptr);  // We don't want to override any existing value
+        auto v = other.get(i);      // pass by reference to denote that we don't own it
```
Comment (perrotta) [inline RecoTracker/TkHitPairs/interface/LayerHitMapCache.h]: Should this comment be moved to L43 now?

File: RecoTracker/TkHitPairs/src/HitPairGeneratorFromLayerPair.cc [hunk 2]
```diff
@@ -152,6 +152,8 @@ void HitPairGeneratorFromLayerPair::doublets(const TrackingRegion& region,
     for (int j = 0; j < 3; j += 2) {
       auto b = innerRange[j];
       auto e = innerRange[j + 1];
+      if (e == b)
```
Comment (perrotta) [inline RecoTracker/TkHitPairs/src/HitPairGeneratorFromLayerPair.cc]: Isn't ```if (e <= b)``` even safer?

Comment (Dr15Jones) [reply RecoTracker/TkHitPairs/src/HitPairGeneratorFromLayerPair.cc]: If b > e then the whole algorithm would be busted, wouldn't it?

File: RecoTracker/TkHitPairs/interface/LayerHitMapCache.h (hunk 1 above)
Comment (Dr15Jones) [reply RecoTracker/TkHitPairs/interface/LayerHitMapCache.h]: done

**Comment by Dr15Jones** [comment]
please test

**Comment by perrotta** [comment]
+1
- It fixes the issues reported in the PR description
- Jenkins tests pass

**Comment by fabiocos** [comment]
+1

**Pull request merged**